src/main/java/org/cryptomator/cryptofs/ch/CleartextFileChannel.java

```diff
@@ -324,18 +325,21 @@ long beginOfChunk(long cleartextPos) {
 	protected void implCloseChannel() throws IOException {
 		try {
 			flush();
-			ciphertextFileChannel.force(true);
-			try {
-				persistLastModified();
-			} catch (NoSuchFileException nsfe) {
-				//no-op, see https://github.com/cryptomator/cryptofs/issues/169
-			} catch (IOException e) {
-				//only best effort attempt
-				LOG.warn("Failed to persist last modified timestamp for encrypted file: {}", e.getMessage());
-			}
 		} finally {
 			super.implCloseChannel();
 			closeListener.closed(this);
+			try {
+				ciphertextFileChannel.close();
+			} finally {
+				try {
+					persistLastModified();
+				} catch (NoSuchFileException nsfe) {
+					//no-op, see https://github.com/cryptomator/cryptofs/issues/169
+				} catch (IOException e) {
+					//only best effort attempt
+					LOG.warn("Failed to persist last modified timestamp for encrypted file: {}", e.getMessage());
+				}
+			}
 		}
 	}
 }
```

Comment:
as per the (badly designed) nio API, `implCloseChannel` is called at most once, regardless of whether closing succeeded or failed. I.e. *everything* we do here must be "best effort" style, there is no chance to re-attempt anything after recovering from an exception.

In other words, all ~four~ five statements must be invoked: `flush()`, `super.implCloseChannel()`, `closeListener.closed(this)`, `ciphertextFileChannel.close()` (+ bonus `persistLastModified()`).

As any of these can throw (e.g. `closeListener.closed(this)`), I believe there is no way around an absurdly nested `try-finally` monster. To keep this readable, I suppose we need to split this up into multiple methods.

Reply:
i split the method into three parts in e247a2e264eb0a6f9e2cff53e77238678d43b842

Reply:
I dislike the names but don't have any better suggestions.

Maybe use a helper function?

```java
protected void implCloseChannel() throws IOException {
	tryAll(List.<CloseAction>of(
		this::flush,
		super::implCloseChannel,
		ciphertextFileChannel::close,
		this::tryPersistLastModified
	).iterator());
}

private void tryAll(Iterator<CloseAction> actions) throws IOException {
	if (actions.hasNext()) {
		try {
			actions.next().run();
		} finally {
			tryAll(actions);
		}
	}
}

@FunctionalInterface
private interface CloseAction {
	void run() throws IOException;
}
```

Reply:
Oh, i like this idea. Keeps the important functions in one place and readable. Changed in 675780129dfacd5cb6f8f0ebcd6ad9b13942eee0